File: app/src/main/java/com/sermage/mymoviecollection/screens/search/SearchViewModel.java
package com.sermage.mymoviecollection.screens.search;

import android.app.Application;
import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import com.sermage.mymoviecollection.api.ApiFactory;
import com.sermage.mymoviecollection.api.ApiService;
import com.sermage.mymoviecollection.data.MovieDatabase;
import com.sermage.mymoviecollection.pojo.Movie;
import com.sermage.mymoviecollection.pojo.MovieResponse;

import java.util.List;
import java.util.Locale;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class SearchViewModel extends AndroidViewModel {

    private MutableLiveData<List<Movie>> searchableMovies;
    private String lang;
    private CompositeDisposable compositeDisposable;
    private static MovieDatabase database;
    private static final String API_KEY="e95df96cf2fe57cc73fe43be0db6c773";

    public SearchViewModel(@NonNull Application application) {
        super(application);
        searchableMovies=new MutableLiveData<>();
        lang= Locale.getDefault().getLanguage();
        database=MovieDatabase.getInstance(application);
    }

    public MutableLiveData<List<Movie>> getSearchableMovies() {
        return searchableMovies;
    }

    public void loadData(String query){
        ApiFactory apiFactory=ApiFactory.getInstance();
        ApiService apiService=apiFactory.getApiService();
        compositeDisposable=new CompositeDisposable();
        Disposable disposable=apiService.getSearchableMovies(API_KEY,lang,query)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<MovieResponse>() {
                    @Override
                    public void accept(MovieResponse response) throws Exception {
                        insertMovies(response.getResults());
                        searchableMovies.setValue(response.getResults());
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {

                    }
                });
        compositeDisposable.add(disposable);
    }

    @Override
    protected void onCleared() {
        compositeDisposable.dispose();
        super.onCleared();
    }

    @SuppressWarnings("unchecked")
    public void insertMovies(List<Movie> movies){
        new InsertMovieTask().execute(movies);
    }

    public static class InsertMovieTask extends AsyncTask<List<Movie>,Void,Void> {

        @SuppressWarnings("unchecked")
        @Override
        protected Void doInBackground(List<Movie>... lists) {
            if(lists!=null && lists.length>0){
                database.movieDao().insertMovies(lists[0]);
            }
            return null;
        }
    }
}

File: app/src/main/java/com/sermage/mymoviecollection/screens/search/SearchViewModel.kt
package com.sermage.mymoviecollection.screens.search

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.sermage.mymoviecollection.api.ApiFactory
import com.sermage.mymoviecollection.pojo.Movie
import com.sermage.mymoviecollection.pojo.TVShow
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class SearchViewModel : ViewModel() {

    companion object{
       var page=1
    }

    private val compositeDisposable = CompositeDisposable()
    private val searchableMovies = MutableLiveData<List<Movie>>()
    private val searchableTvShows = MutableLiveData<List<TVShow>>()
    private val moviesByGenre=MutableLiveData<List<Movie>>()
    private val tvShowsByGenre=MutableLiveData<List<TVShow>>()
    private var isLoadingMovies=MutableLiveData<Boolean>()
    private var isLoadingTVShow=MutableLiveData<Boolean>()

    private var searchableMoviesList= mutableListOf<Movie>()
    private var searchableTVShowList= mutableListOf<TVShow>()
    private var moviesByGenreList= mutableListOf<Movie>()
    private var tvShowByGenreList= mutableListOf<TVShow>()


    fun getSearchableMovies(): MutableLiveData<List<Movie>> {
        return searchableMovies
    }

    fun getSearchableTvShows(): LiveData<List<TVShow>> {
        return searchableTvShows
    }

    fun getMoviesByGenre():LiveData<List<Movie>>{
        return moviesByGenre
    }

    fun getTVShowsByGenre():LiveData<List<TVShow>>{
        return tvShowsByGenre
    }

    fun getStatusOfLoadingMovies():LiveData<Boolean>{
        return isLoadingMovies
    }
    fun getStatusOfLoadingTV():LiveData<Boolean>{
        return isLoadingTVShow
    }

    fun getSearchingMoviesList():MutableList<Movie>{
        return searchableMoviesList
    }
    fun getSearchingTVShowList():MutableList<TVShow>{
        return searchableTVShowList
    }

    fun loadMovies(query: String) {
        val loading = isLoadingMovies.value
                if (loading != null && loading) {
                    return
                }
        isLoadingMovies.postValue(true)
        val disposable = ApiFactory.apiService.getSearchableMovies(query = query,page =page)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                isLoadingMovies.postValue(false)
                it.results?.let { it1 -> searchableMoviesList.addAll(it1) }
                searchableMovies.value = searchableMoviesList
                page++
            }, {
                isLoadingMovies.postValue(false)
            })
        compositeDisposable.add(disposable)

    }

    fun loadTvShows(query: String) {
        val loading = isLoadingTVShow.value
        if (loading != null && loading) {
            return
        }
        isLoadingTVShow.postValue(true)
        val disposable = ApiFactory.apiService.getSearchableTVShows(query = query,page = page)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                isLoadingTVShow.postValue(false)
                it.results?.let { it1 -> searchableTVShowList.addAll(it1) }
                searchableTvShows.value = searchableTVShowList
                page++
            }, {
                isLoadingTVShow.postValue(false)
            })
        compositeDisposable.add(disposable)

    }

    fun loadMoviesByGenre(genre:String){
        val disposable=ApiFactory.apiService.getMoviesByGenre(genres=genre,page=page)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                it.results?.let { it1 -> moviesByGenreList.addAll(it1) }
                moviesByGenre.value=moviesByGenreList
                page++
            },{

            })
        compositeDisposable.add(disposable)
    }

    fun loadTVShowsByGenre(genre:String){
        val disposable=ApiFactory.apiService.getTVShowsByGenre(genres=genre,page=page)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                it.results?.let { it1 -> tvShowByGenreList.addAll(it1) }
                tvShowsByGenre.value=tvShowByGenreList
                page++
            },{

            })
        compositeDisposable.add(disposable)
    }


    override fun onCleared() {
        super.onCleared()
        compositeDisposable.dispose()
    }
}